Assets/MixedRealityToolkit.Services/InputSystem/FocusProvider.cs

```diff
@@ -412,6 +412,97 @@ public override int GetHashCode()
             }
         }
 
+        /// <summary>
+        /// Helper class for managing the visibility of the gaze pointer to match windows mixed reality and HoloLens 2
+        /// When application starts, gaze pointer is visible. Then when articulate hands / motion controllers
+        /// appear, hide the gaze cursor. Whenever user says the voice wake word, make the 
+        /// gaze controller appear.
+        /// </summary>
+        private class GazePointerStateMachine : IMixedRealitySpeechHandler
+        {
+            private enum GazePointerState
+            {
+                Initial, // When the application starts up, the gaze pointer should be active
+                GazePointerActive, // Gaze pointer is active when no hands are visible, after "select" 
+                GazePointerInactive // Gaze pointer is inactive as soon as motion controller or articulated hand pointers appear
+            }
+            private GazePointerState gazePointerState = GazePointerState.Initial;
+            private bool activateGazeKeywordIsSet = false;
+
+            public bool IsGazePointerActive
+            {
+                get { return gazePointerState != GazePointerState.GazePointerInactive; }
+            }
+
+            public void UpdateState(HashSet<PointerData> pointers, GenericPointer currentGazePointer)
+            {
+                int numFarPointers = 0;
+                int numNearPointersActive = 0;
+                foreach (var p in pointers)
+                {
+                    if (p.Pointer != null)
+                    {
+                        if (p.Pointer is IMixedRealityNearPointer)
+                        {
+                            if (p.Pointer.IsInteractionEnabled)
```

Comment:
You could reduce nesting by doing and && here with the condition above.

Reply:
I could, but then the meaning of the else clause below would mean "it is not near and enabled", meaning that if the pointer was NearPointer and disabled, that it would fall into the else clause. Even though it might be okay, I think I'd prefer to keep it this way for now, just so it's more clear what what else clause captures.